sdk/src/main/java/com/google/cloud/dataflow/sdk/io/BoundedSource.java

```diff
@@ -133,6 +133,96 @@ public Double getFractionConsumed() {
     }
 
     /**
+     * A constant to use as the return value for {@link #getParallelismConsumed()} or
+     * {@link #getParallelismRemaining()} when the exact value is unknown.
+     */
+    public static final long PARALLELISM_UNKNOWN = -1;
+
+    /**
+     * Returns the total amount of parallelism in the consumed (and processed) range of this
+     * reader's current {@link BoundedSource} (as would be returned by {@link #getCurrentSource}).
+     *
+     * <p>The "parallelism" consumed does not necessarily equal the number of <em>records</em>
+     * output and processed, but rather the number of distinct (ideally non-empty) <em>sources</em>
+     * that could be produced by calls to {@link #splitAtFraction}.
+     *
+     * <p>Consider the following two examples. An input that can be read in parallel down to the
+     * individual records, such as {@link CountingSource#upTo}, is called "perfectly splittable".
+     * Conversely, an example of a non-perfectly splittable input is a block-compressed file format
+     * such as {@link AvroIO}, in which a block of records has to be read as a whole, but different
+     * blocks can be read in parallel.
+     *
+     * <ul>
+     * <li>Any {@link BoundedReader reader} that is unstarted (aka, has never had a call to
+     * {@link #start}) has a consumed parallelism of 0.
+     * <li>Any {@link BoundedReader reader} that is currently processing its first element (aka,
```

Comment:
This is a bit confusing because the reader is not _processing_ the element. Better just say "that has not yet returned its second element" and remove "and thus has not been consumed".